Add `elastic-agent-wolfi-complete` Docker image

rdner opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->
The PR adds a new Wolfi-based `elastic-agent-wolfi-complete` Docker image that contains Synthetics. 

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->
It's useful, for example, if a configured integration is running Heartbeat. It also completes the migration of all images to Wolfi.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

```sh
PLATFORMS=linux/arm64 PACKAGES=docker mage package
```

Before:

```
REPOSITORY                                       TAG          IMAGE ID       CREATED         SIZE
docker.elastic.co/beats-ci/elastic-agent-cloud   8.16.0       3d78b08a1856   6 minutes ago   939MB
docker.elastic.co/beats-dev/golang-crossbuild    1.22.6-arm   eae46a040730   11 days ago     752MB
docker.elastic.co/beats/elastic-agent            8.16.0       c522ff44933d   6 minutes ago   647MB
docker.elastic.co/beats/elastic-agent-complete   8.16.0       0f72cb8295c2   5 minutes ago   2.16GB
docker.elastic.co/beats/elastic-agent-ubi        8.16.0       09c24222756b   6 minutes ago   644MB
docker.elastic.co/beats/elastic-agent-wolfi      8.16.0       48fe4922082a   6 minutes ago   573MB
```

After:

```
REPOSITORY                                             TAG          IMAGE ID       CREATED         SIZE
docker.elastic.co/beats-ci/elastic-agent-cloud         8.16.0       8d38473ba140   9 minutes ago   939MB
docker.elastic.co/beats-dev/golang-crossbuild          1.22.6-arm   eae46a040730   11 days ago     752MB
docker.elastic.co/beats/elastic-agent                  8.16.0       734a01ade09b   9 minutes ago   647MB
docker.elastic.co/beats/elastic-agent-complete         8.16.0       fadd98e9ca69   8 minutes ago   2.16GB
docker.elastic.co/beats/elastic-agent-ubi              8.16.0       d80f9c55d596   9 minutes ago   644MB
docker.elastic.co/beats/elastic-agent-wolfi            8.16.0       fc91b6e99adc   9 minutes ago   573MB
docker.elastic.co/beats/elastic-agent-wolfi-complete   8.16.0       f7bf606c8fea   7 minutes ago   1.96GB
```

I enrolled an agent running on the new `elastic-agent-wolfi-complete` image with the flags from here https://www.elastic.co/guide/en/observability/current/synthetics-private-location.html#synthetics-private-location-connect:
<img width="1712" alt="Screenshot 2024-09-10 at 16 44 02" src="https://github.com/user-attachments/assets/cc0f1661-314a-4360-85bd-8b68a0e973ca">

Created a private location in Synthetics:
<img width="1659" alt="Screenshot 2024-09-10 at 16 44 13" src="https://github.com/user-attachments/assets/a6cf64a0-3e84-4b57-9f62-309592894ef3">

I created a lightweight monitor (HTTP ping) for https://www.elastic.co/docs/current/integrations and a browser monitor (multistep) for this script:

```javascript
step('Load the System integration page', async () => {
  await page.goto('https://www.elastic.co/docs/current/integrations');
});
```

Both monitors succeeded:

<img width="1671" alt="Screenshot 2024-09-11 at 10 12 12" src="https://github.com/user-attachments/assets/b77e783b-069c-47d9-b5b7-f25197ce2921">

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Relates to https://github.com/elastic/elastic-agent/pull/5062
- Relates to https://github.com/elastic/beats/pull/40689

<!-- CI Cheatsheet
Trigger comments:
/test             (Or `buildkite test this|it`) Triggers unit test pipeline
/test extended    (Or `buildkite test extended`) Triggers integration test pipeline

PR labels:
skip-ci           Skips unit and integration tests
skip-it           Skips integration tests
-->

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane (Team:Elastic-Agent-Control-Plane)

blakerouse reviewed (changes requested):
This looks good and I don't see any issues with this PR, except for all the switches of `'`' to `"`. That really pollutes this PR and when doing a git show once this is merged it is going to show many line changes that are unrelated to creating `wolfi-complete`. Could you please split out the `'` to `"` change to its own PR?

rdner commented:
@blakerouse sorry about that, I'll clean it up.

cmacknz commented:
How'd we test this? Can you create a Synethics private location using the agent container? https://www.elastic.co/guide/en/observability/current/synthetics-settings.html#synthetics-settings-private-locations

CC @emilioalvap @vigneshshanmugam

vigneshshanmugam commented on dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:
```diff
@@ -239,8 +239,36 @@ RUN for iter in {1..10}; do \
+    {{ $beatHome }}/.synthetics \
+    {{ $beatHome }}/.npm \
+    {{ $beatHome }}/.cache \
+    | xargs -IDIR sh -c 'mkdir -m 0770 -p DIR && chown -R {{ .user }} DIR'
```
If we can keep the permission in sync with the other PR - https://github.com/elastic/beats/pull/40689/files#r1750771988

rdner replied:
Sorry, I have not seen the original comment https://github.com/elastic/beats/pull/40689/files#r1750771988 and thought it was a typo which I "fixed" in https://github.com/elastic/beats/commit/b06f7cea66b37978bc985a04aa851b6ea6fc2519

I think I need more context on how switching permissions `0770` to `0775` reflects removing the agent from the root group. The user created for this container is the owner of the entire directory tree and the last `0` in the bitmask denies access to "other" users and groups. Why should it be "read and execute" instead?

rdner replied:
I think @emilioalvap would be the right person to ask in this case 🙂

vigneshshanmugam commented on dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:
```diff
@@ -239,8 +239,36 @@ RUN for iter in {1..10}; do \
+# Install required dependencies from wolfi repository
+RUN for iter in {1..10}; do \
+        apk update && \
+        apk add --no-interactive --no-progress --no-cache nodejs-18=18.20.4-r0 npm=10.8.3-r0 glib dbus-glib libatk-1.0 \
```
can you update the NODE_VERSION on ubuntu as wel - https://github.com/elastic/elastic-agent/blob/cec25d6e8fb0b76f5a50e78d022d4fcffad26287/dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl#L195

rdner replied:
done in https://github.com/elastic/elastic-agent/pull/5475/commits/6d9ad70957ae24046997911cfbecf004dca8f0cf

vigneshshanmugam reviewed (approved):
Changes looks good to me. 

I would test this by packaging and adding it as Synthetics private location as @cmacknz commented out. Steps would be

1. Package it up and run the [elastic agent with specific flags ](https://www.elastic.co/guide/en/observability/current/synthetics-private-location.html#synthetics-private-location-connect)2. Start the Kibana and add the agent as [Private location](https://www.elastic.co/guide/en/observability/current/synthetics-private-location.html#synthetics-private-location-add)
3. Now you go the Synthetics App and create [both Lightweight and browser monitors](https://www.elastic.co/guide/en/observability/current/synthetics-get-started-ui.html) to confirm they are running as expected

rdner commented:
I'll go through the testing steps from https://github.com/elastic/elastic-agent/pull/5475#pullrequestreview-2291160553 and will update the description of the PR once done.

rdner commented:
@vigneshshanmugam @emilioalvap I might need some help here.

I've created a browser (muiltistep) monitor with the following script:

```javascript
step('Load the System integration page', async () => {
  await page.goto('https://www.elastic.co/docs/current/integrations');
});
```

And I get this error on its execution:

```
⚠ Throttling may not be active when the tests run - see
https://github.com/elastic/synthetics/blob/main/docs/throttling.md for more details
browserType.launch: Executable doesn't exist at /usr/share/elastic-agent/.cache/ms-playwright/chromium-1124/chrome-linux/chrome
╔═════════════════════════════════════════════════════════════════════════╗
║ Looks like Playwright Test or Playwright was just installed or updated. ║
║ Please run the following command to download new browsers: ║
║ ║
║ npx playwright install ║
║ ║
║ <3 Playwright Team ║
╚═════════════════════════════════════════════════════════════════════════╝
at Function.setupDriver (/usr/share/elastic-agent/.npm/lib/node_modules/@elastic/synthetics/src/core/gatherer.ts:56:43)
at Function.createContext (/usr/share/elastic-agent/.npm/lib/node_modules/@elastic/synthetics/src/core/runner.ts:83:35)
at Runner.runJourney (/usr/share/elastic-agent/.npm/lib/node_modules/@elastic/synthetics/src/core/runner.ts:367:34)
at Runner.run (/usr/share/elastic-agent/.npm/lib/node_modules/@elastic/synthetics/src/core/runner.ts:496:22)
at Command.<anonymous> (/usr/share/elastic-agent/.npm/lib/node_modules/@elastic/synthetics/src/cli.ts:143:23) {
name: 'Error'
}
```

And, indeed, I don't see Chromium installation in the original PR for Heartbeat https://github.com/elastic/beats/pull/40689

Unlike the Ubuntu section here https://github.com/elastic/beats/blob/88c84a09ef35f4c5c5548562f7e1b796eec8fe8f/dev-tools/packaging/templates/docker/Dockerfile.tmpl#L217 the new code for Wolfi does not have Chromium.

I see that the Heartbeat PR was well-tested (see description) and I wonder how it worked if the same setup with Elastic Agent does not? 

Can you help me with the missing parts of the installation script?

vigneshshanmugam commented:
@rdner Good catch, Indeed we dont have the deps. I was thinking the installation of synthetics would auto-install the playwright browsers but this has been changed recently. We need to fix also on the HB layer.

Can you add this to the wolfi image
```
$NODE_PATH/node/lib/node_modules/@elastic/synthetics/node_modules/.bin/playwright install chromium
```

This should install the chromium browser that we need to run the browser monitors.

rdner commented:
@vigneshshanmugam if I add this line I get:

```
 => ERROR [stage-1 14/15] RUN $NODE_PATH/node/lib/node_modules/@elastic/synthetics/node_modules/.bin/playwright install chromium   0.3s
------                                                                                                                                  
 > [stage-1 14/15] RUN $NODE_PATH/node/lib/node_modules/@elastic/synthetics/node_modules/.bin/playwright install chromium:
0.107 /bin/sh: /node/lib/node_modules/@elastic/synthetics/node_modules/.bin/playwright: not found
------
```

Looks like we need to install `playwright` manually as well.

vigneshshanmugam commented:
@rdner My bad, the issue here is that `@elastic/synthetics` is installed on a different directory than the one I sent (it was on ubuntu). If you can exec in to the container and find the NPM bin paths, we can get the correct path by appending the relevant bits. 

```
which elastic-synthetics // will give the relevant bin path for the Elastic synthetics library

// replace the bin path with lib and invoke the Playwright cli
<extacted path till .bin>/lib/node_modules/@elastic/synthetics/node_modules/.bin/playwright
```

rdner commented:
@emilioalvap @vigneshshanmugam my bad, in the Dockerfile I switched to the root user before installing synthetics and all the dependencies got installed in a wrong directory.

Fixed it in https://github.com/elastic/elastic-agent/pull/5475/commits/0f354a76b7813e83100c69e671b213c63128a2fa and everything worked as expected (see the description of this PR).
